java/build/bazel/tests/integration/BazelCommand.java

```diff
@@ -0,0 +1,148 @@
+// limitations under the License.
+
+package build.bazel.tests.integration;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/** This class holds the result of a Bazel invocation. */
+public class BazelCommand {
+
+  private final Command delegate;
+  private final List<String> args;
+  private final int exitCode;
+
+  private BazelCommand(Command delegate, List<String> args, int exitCode) {
+    this.delegate = delegate;
+    this.args = args;
+    this.exitCode = exitCode;
+  }
+
+  /** Returns the exit code. */
+  public int exitCode() {
+    return exitCode;
+  }
+
+  /** Returns the list of lines of the standard error stream. */
+  public List<String> errorLines() {
+    return delegate.getErrorLines();
+  }
+
+  /** Returns the list of lines of the standard output stream. */
+  public List<String> outputLines() {
+    return delegate.getOutputLines();
+  }
+
+  /**
+   * Returns a description/summary of the command arguments, standard error and output stream. This
+   * method can be used for troubleshooting and error reporting.
+   */
+  public String toString() {
+    String description =
+        "BAZEL COMMAND: "
+            + args
+            + "\n"
+            + "EXIT CODE: "
+            + exitCode
+            + "\n"
+            + "STDOUT:\n    "
+            + String.join("\n    ", outputLines())
+            + "\n"
+            + "STDERR:\n    "
+            + String.join("\n    ", errorLines())
+            + "\n";
+
+    return description;
+  }
+
+  static class Builder {
+    private final WorkspaceDriver driver;
+    private final Path outputUserRoot;
+    private final List<String> args;
+    private Path bazelrcFile = null;
+    private Map<String, String> environment = new HashMap<>();
+    private Path workingDirectory = Paths.get("");
+
+    Builder(WorkspaceDriver driver, Path outputUserRoot, List<String> args) {
+      this.driver = driver;
+      this.outputUserRoot = outputUserRoot;
+      this.args = Collections.unmodifiableList(args);
+    }
+
+    /** sets the working directory relative to workspace path */
+    public Builder inWorkingDirectory(Path dirRelativeToWorkspaceRoot) {
+      this.workingDirectory = dirRelativeToWorkspaceRoot;
+      return this;
+    }
+
+    /** sets the bazelrc file path */
+    public Builder withBazelrcFile(Path bazelrcFile) {
+      this.bazelrcFile = bazelrcFile;
+      return this;
+    }
+
+    /** adds environment variable to execution runtime */
+    public Builder withEnvironmentVariable(String key, String value) {
+      this.environment.put(key, value);
+      return this;
+    }
+
+    private Command build() throws IOException {
+      String bazelRcPath =
+          bazelrcFile == null
```

Comment:
why do you prefer null and trenary over optional?

Reply:
The (official) position that I am following is that Optional in Java was not intended for class fields. Of course, this is a matter of opinion and coding style, and you are the final arbiter :)

Cf. answer by Brian Goetz: https://stackoverflow.com/questions/26327957/should-java-8-getters-return-optional-type/26328555#26328555 and also http://www.oracle.com/technetwork/articles/java/java8-optional-2175753.html: "The purpose of Optional is not to replace every single null reference in your codebase but rather to help design better APIs in which—just by reading the signature of a method—users can tell whether to expect an optional value."

Null is too much embedded in the Java language for using Optional in all cases.  Optional was primarily intended for method chaining and for public APIs, especially when they relate to streams.  IntelliJ agrees, and flags the use of Optional for fields.

It is also possible to use the @Nullable annotation to help findbugs finds spurious usage (findbugs is enabled by default in Bazel's java toolchain).

Reply:
Ok. I’ll live with it though I’m glad I don’t code in java in the d2d...